src/Acrobot.jl

```diff
@@ -0,0 +1,222 @@
+               William Dabney, Jonathan P. How
+License: "BSD 3-Clause"
+Author: "Christoph Dann <cdann@cdann.de>"
+"""
+mutable struct AcrobotEnv
+    viewer
+    observation_space::Box
+    action_space::Discrete
+    state
+    dt::Float32
+    LINK_LENGTH_1::Float32  # [m]
+    LINK_LENGTH_2::Float32  # [m]
+    LINK_MASS_1::Float32  # [kg] mass of link 1
+    LINK_MASS_2::Float32  # [kg] mass of link 2
+    LINK_COM_POS_1::Float32  # [m] position of the center of mass of link 1
+    LINK_COM_POS_2::Float32  # [m] position of the center of mass of link 2
+    LINK_MOI::Float32  # moments of intertia for both links
+    MAX_VEL_1::Float32
+    MAX_VEL_2::Float32
+    AVAIL_TORQUE::Array{Float32, N} where N
+    torque_noise_max::Float32
+    book_or_nips::String
+end
+
+
+function AcrobotEnv()
+    dt = 2f-1
+    LINK_LENGTH_1 = 1f0
+    LINK_LENGTH_2 = 1f0
+    LINK_MASS_1 = 1f0
+    LINK_MASS_2 = 1f0
+    LINK_COM_POS_1 = 5f-1
+    LINK_COM_POS_2 = 5f-1
+    LINK_MOI = 1f0
+
+    MAX_VEL_1 = Float32(4 * pi)
+    MAX_VEL_2 = Float32(9 * pi)
+
+    AVAIL_TORQUE = [-1f0, 0f0, 1f0]
+    torque_noise_max = 0f0
+
+    # use dynamics equations from the nips paper or the book
+    book_or_nips = "book"
+    # action_arrow = nothing
+    # domain_fig = nothing
+    # actions_num = 3
+
+    viewer = nothing
+    high = [1f0, 1f0, 1f0, 1f0, MAX_VEL_1, MAX_VEL_2] #//
+    low = -high
+    observation_space = Box(low, high, Float32)
+    action_space = Discrete(3)
+    state = nothing
+
+    AcrobotEnv(viewer, observation_space, action_space, state, dt, LINK_LENGTH_1,
+               LINK_LENGTH_2, LINK_MASS_1, LINK_MASS_2, LINK_COM_POS_1, LINK_COM_POS_2,
+               LINK_MOI, MAX_VEL_1, MAX_VEL_2, AVAIL_TORQUE, torque_noise_max, book_or_nips)
+end
+
+
+function reset!(env::AcrobotEnv)
+    env.state = param(rand(Float32, 4) * 2f-1 .- 1f-1)
+    return _get_ob(env)
+end
+
+#=
+function custom_reset!(env::AcrobotEnv, args=[-0.04616826f0, -0.04653378f0,  0.07321809f0, -0.04461966f0])
+    env.state = param(args)
+    return _get_ob(env)
+end
+=#
+get_torque(env::AcrobotEnv, action) = env.AVAIL_TORQUE[action]
+
+function step!(env::AcrobotEnv, action)
+    @assert action ∈ env.action_space "$(action) is not in the action space. Enter a valid action."
+    s = env.state
+    torque = get_torque(env, action)
+
+    # Add noise to the force action
+    env.torque_noise_max > 0 && (torque += env.torque_noise_max * (2rand(Float32) - 1))
+
+    # Now, augment the state with our force action so it can be passed to _dsdt
+    s_augmented = vcat(s, torque)
+
+
+    ns = rk4(env, _dsdt, s_augmented, [0, env.dt])
+    # only care about final timestep of integration returned by integrator
+    ns = ns[end, :]
+    ns = ns[1:4]  # omit action
+    ns.data[1] = wrap(ns[1], -pi, pi).data
+    ns.data[2] = wrap(ns[2], -pi, pi).data
+    ns.data[3] = bound(ns[3], -env.MAX_VEL_1, env.MAX_VEL_2).data
+    ns.data[4] = bound(ns[4], -env.MAX_VEL_1, env.MAX_VEL_2).data
```

Comment:
Why is only `data` being changed here? Curious to know the consequences if we rather treat it as a vector of one element and perform operations on it in `wrap` and `bound`.

Reply:
In its current state, both `wrap` and `bound` return a `TrackedReal`. Even though `ns[1]`, `ns[2]` etc are elements of a `TrackedArray`, directly setting `ns[1] = x` where `x` Is tracked, gives a `cannot differentiate 'setindex!'` error. I asked around in the community, and I was informed that in-place mutations of elements of tracked arrays aren't currently supported. Is there any other way of achieving what you're looking for?